src/v3/core/graph.test.js

```diff
@@ -96,14 +122,207 @@ describe("core/graph", () => {
+              .addNode(src)
+              .addNode(dst)
+              .addEdge({src, dst, address});
+          it("`hasEdge` can discover the edge", () => {
+            expect(graph().hasEdge(address)).toBe(true);
+          });
+          it("`edge` can retrieve the edge", () => {
+            expect(graph().edge(address)).toEqual({src, dst, address});
+          });
+          it("`edges` contains the edge", () => {
+            const edgeArray = (g: Graph) => Array.from(g.edges());
+            expect(edgeArray(graph())).toEqual([{src, dst, address}]);
+          });
+        });
+
+        describe("with edge added and removed", () => {
+          const graph = () =>
+            new Graph()
+              .addNode(src)
+              .addNode(dst)
+              .addEdge({src, dst, address})
+              .removeEdge(address);
+          it("`hasEdge` now returns false", () => {
+            expect(graph().hasEdge(address)).toBe(false);
+          });
+          it("`edge` returns undefined", () => {
+            expect(graph().edge(address)).toBe(undefined);
+          });
+          it("`edges` is empty", () => {
+            expect(edgeArray(graph())).toHaveLength(0);
+          });
+          it("nodes were not removed", () => {
+            expect(graph().hasNode(src)).toBe(true);
+            expect(graph().hasNode(dst)).toBe(true);
+            expect(Array.from(graph().nodes())).toHaveLength(2);
+          });
+        });
+
+        describe("with multiple loop edges", () => {
+          const e1 = edgeAddress(["e1"]);
+          const e2 = edgeAddress(["e2"]);
+          const edge1 = {src, dst: src, address: e1};
+          const edge2 = {src, dst: src, address: e2};
+          const quiver = () =>
+            new Graph()
+              .addNode(src)
+              .addEdge(edge1)
+              .addEdge(edge2);
+          it("adding multiple loop edges throws no error", () => {
+            quiver();
+          });
+          it("both edges are discoverable via `hasEdge`", () => {
+            expect(quiver().hasEdge(e1)).toBe(true);
+            expect(quiver().hasEdge(e2)).toBe(true);
+          });
+          it("both edges are retrievable via `edge`", () => {
+            expect(quiver().edge(e1)).toEqual(edge1);
+            expect(quiver().edge(e2)).toEqual(edge2);
+          });
+          it("both edges are retrievable from `edges`", () => {
+            expect(edgeArray(quiver()).sort()).toEqual([edge1, edge2].sort());
+          });
+        });
+      });
+
+      describe("idempotency of", () => {
+        const src = nodeAddress(["src"]);
+        const dst = nodeAddress(["dst"]);
+        const address = edgeAddress(["hi"]);
+        it("`addEdge`", () => {
+          const g = new Graph()
+            .addNode(src)
+            .addNode(dst)
+            .addEdge({src, dst, address})
+            .addEdge({src, dst, address});
+          expect(edgeArray(g)).toEqual([{src, dst, address}]);
+        });
```

Comment:
One thing that we don’t test here is that if you add a node twice, then
remove it once, the node is actually removed (not “buffered”). That’s
probably okay to omit. Thoughts?

Reply:
Yeah, I think it's fine.